### ISSUE TITLE
Add a setting to parse crlf with Tab separated CSV (TSV?!) files

### ISSUE THREAD

**Issue by mneedham** [ClickHouse/ClickHouse#56257]
Not sure if this is a feature or usability, so feel free to re-label. 

I have a tab separated file with windows line endings, which means it doesn't work correctly with any of the TSV formats. It will process with the CSV ones:

```
FROM file(
     'data/Stock Histories.zip :: {NASDAQ,NYSE}.txt', 
     CSVWithNames
)
SELECT *, _file
ORDER BY rand()
LIMIT 10;
```

But then every field is inside one row and I have to do a split on tab to separate them.

@alexey-milovidov said maybe we could add a setting:

> TSV does not have quoting, and escaping of \r is not mandatory, so the interpretation of crlf will be ambiguous if you have a binary string.
> We can add it under a setting…

**Comment by Blargian** [comment]
@Avogar I'll work on this one

**Comment by Avogar** [comment]
> The intention with this is that if the added setting is true then we will parse TSV files with line ending \r\n?

Yes. When new setting is true, we should be able to parse files with line ending '\r\n' correctly.

> Does any functionality already exist to deal with a return carriage from a ReadBuffer & buf?

We can work with `\r` similar how we work with `\n` in existing code. Let me explain possible solution:
To read values in TSV format we sometimes need to read all data until `\t` or `\n`, and in all these cases we should support `\r\n` under a setting. For reading TSV values we use `ISerialization::deserializeTextEscaped/deserializeTextRaw` functions (*Raw function for TSVRaw format with no escaping):
https://github.com/ClickHouse/ClickHouse/blob/ca124df92b69925d230c130a5347dcd7e8b7facc/src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp#L181
https://github.com/ClickHouse/ClickHouse/blob/ca124df92b69925d230c130a5347dcd7e8b7facc/src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp#L168-L182

So, we should go through all `deserializeTextEscaped/deserializeTextRaw` implementations and add support '\r' when needed under a setting. We also have functions `readEscapedString/readEscapedStringInto/readTSVField` that are used for reading fields in TSV format, we should support `\r\n` there too, maybe we can just make these functions template with parameter `bool support_crlf`, all these functions are implemented using `readEscapedStringIntoImpl`, so we can propogate this template parameter there and add support for `\r\n` under `if constexpr (support_crlf)`.

**Comment by Blargian** [comment]
@Avogar I've gone through the functions you mentioned above and added support for `\r\n` on a setting. One thing I'm still trying to figure out is that here

https://github.com/ClickHouse/ClickHouse/blob/32d11cdc2da8dc673968194001afcccc8c630104/src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp#L140-L152

`checkForCarriageReturn(*buf)` gets called inside `skipRowEndDelimiter()`. Now that I modified `readEscapedStringIntoImpl` the fields get read in correctly without `\r` depending on if setting is true or not. As I see it `ReadBuffer` object doesn't get modified. I mean that we don't remove `\r\n` from this buffer, we just have logic to skip the delimiters (by advancing position of ReadBuffer) and put characters before this delimiter into the Field. However `checkForCarriageReturn` is still checking this unmodified `ReadBuffer`:

https://github.com/ClickHouse/ClickHouse/blob/32d11cdc2da8dc673968194001afcccc8c630104/src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp#L25-L32 

So we will need to modify this function as well not to check in `ReadBuffer` for \r at the end of a line based on if setting value is true?
